starlite/contrib/sqlalchemy/repository.py

```diff
@@ -0,0 +1,270 @@
+"""SQLAlchemy-based implementation of the repository protocol."""
+from __future__ import annotations
+
+from contextlib import contextmanager
+from typing import TYPE_CHECKING, Any, Generic, Literal, TypeVar
+
+from sqlalchemy import select, text
+from sqlalchemy.exc import IntegrityError, SQLAlchemyError
+
+from starlite.contrib.repository.abc import AbstractRepository
+from starlite.contrib.repository.exceptions import ConflictError, RepositoryError
+from starlite.contrib.repository.filters import (
+    BeforeAfter,
+    CollectionFilter,
+    LimitOffset,
+)
+
+if TYPE_CHECKING:
+    from collections import abc
+    from datetime import datetime
+
+    from sqlalchemy import Select
+    from sqlalchemy.engine import Result
+    from sqlalchemy.ext.asyncio import AsyncSession
+
+    from starlite.contrib.repository.types import FilterTypes
+    from starlite.contrib.sqlalchemy import base
+
+__all__ = [
+    "SQLAlchemyRepository",
+    "ModelT",
+]
+
+T = TypeVar("T")
+ModelT = TypeVar("ModelT", bound="base.Base | base.AuditBase")
+SQLARepoT = TypeVar("SQLARepoT", bound="SQLAlchemyRepository")
+
+
+@contextmanager
+def wrap_sqlalchemy_exception() -> Any:
+    """Do something within context to raise a `RepositoryError` chained
+    from an original `SQLAlchemyError`.
+
+        >>> try:
+        ...     with wrap_sqlalchemy_exception():
+        ...         raise SQLAlchemyError("Original Exception")
+        ... except RepositoryError as exc:
+        ...     print(f"caught repository exception from {type(exc.__context__)}")
+        ...
+        caught repository exception from <class 'sqlalchemy.exc.SQLAlchemyError'>
+    """
+    try:
+        yield
+    except IntegrityError as exc:
+        raise ConflictError from exc
+    except SQLAlchemyError as exc:
+        raise RepositoryError(f"An exception occurred: {exc}") from exc
+
+
+class SQLAlchemyRepository(AbstractRepository[ModelT], Generic[ModelT]):
+    """SQLAlchemy based implementation of the repository interface."""
+
+    def __init__(self, *, session: AsyncSession, select_: Select[tuple[ModelT]] | None = None, **kwargs: Any) -> None:
```

Comment:
Shouldn't this be `tuple[ModelT, ...]?

Reply:
Yes, nice catch: https://github.com/sqlalchemy/sqlalchemy/blob/9d9e47fe9398b6bcda917f63441d2b5ad4b345c4/lib/sqlalchemy/sql/_typing.py#L155

Reply:
Actually, no, it is `tuple[ModelT]`, that type is the type of row returned from the select query. When making an orm select, the query returns rows like `(Model,)`, and if a query returns more than one row the response is `[(Model,), (Model,)]`.
This is what the `scalar()`, `scalar_one()`, `scalar_first()` accessor methods are for, to return the first element of the row.

Reply:
Ok... Weird